Improve output of jhipster upgrade

gmarziou opened the pull request:
This will ease debugging `jhipster upgrade` issues.

- Use log.ok() standard method in success() logging method
- Skip install if there are conflicts in `package.json` file
- List the conflicting files  at the end if any
- Set verbose output by default
- Add a `--silent`option

Fix #6237

- Please make sure the below checklist is followed for Pull Requests.

- [X] [Travis tests](https://travis-ci.org/jhipster/generator-jhipster/pull_requests) are green
- [ ] Tests are added where necessary
- [X] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

tientq commented:
That's much better, awesome.

deepu105 commented on cli/jhipster.js:
```diff
@@ -35,6 +35,10 @@ if (!semver.satisfies(currentNodeVersion, minimumNodeVersion)) {
+/* WARNING: due to this code, it's a bit more difficult to test when developing
+ * the upgrade generator, workaround consists in running it this way:
+ * node generator-jhipster/cli/cli.js upgrade
+ */
```
should we add a flag to control this behavior like `--use-local` and `--use-global` which can override these

gmarziou replied:
Yes but I must test that it addresses this particular use case when developing the generator because unfortunately, yarn does not handle binaries well when using link.
I'll give feedback on this.

deepu105 commented on generators/upgrade/index.js:
```diff
@@ -82,22 +84,24 @@ module.exports = UpgradeGenerator.extend({
-        shelljs.exec(`${generatorCommand} --with-entities --force --skip-install`, { silent: this.silent }, (code, msg, err) => {
-            if (code === 0) this.log(chalk.green(`Successfully regenerated application with JHipster ${version}`));
+        const regenerateCmd = `${generatorCommand} --with-entities --force --skip-install`;
+        this.log.info(regenerateCmd);
```
shouldnt this be `this.info`?

gmarziou replied:
No because this.log.info() provides 2 things I like here: indentation and grey info prefix like you can see in [screenshot](https://github.com/jhipster/generator-jhipster/issues/6237#issuecomment-323235768).

deepu105 replied:
Yes but your info method also calls the same `this.log.info` rite?
```
+    info(msg) {
 +        this.log.info(msg);
 +    }
```

gmarziou replied:
Right, I don't remember, I probably forgot about it. I'll have a look.

gmarziou replied:
OK just checked, you're right. I replaced all calls to `this.log.info()` by `this.info()`

deepu105 commented on generators/upgrade/index.js:
```diff
@@ -106,12 +110,15 @@ module.exports = UpgradeGenerator.extend({
     _regenerate(version, callback) {
         this._generate(version, () => {
             if (this.clientFramework === 'angular1' && version === this.latestVersion) {
+                this.log.info('bower install');
```
same here

deepu105 commented on generators/upgrade/index.js:
```diff
@@ -106,12 +110,15 @@ module.exports = UpgradeGenerator.extend({
             }
-            shelljs.rm('-Rf', `${SERVER_MAIN_RES_DIR}keystore.jks`);
+            const keystore = `${SERVER_MAIN_RES_DIR}keystore.jks`;
+            this.log.info(`Removing ${keystore}`);
```
here as well

deepu105 commented on generators/upgrade/index.js:
```diff
@@ -225,8 +232,10 @@ module.exports = UpgradeGenerator.extend({
-                shelljs.exec(`${commandPrefix} ${GENERATOR_JHIPSTER}@${version} --dev --no-lockfile --ignore-scripts`, { silent: this.silent }, (code, msg, err) => {
-                    if (code === 0) this.log(chalk.green(`Installed ${GENERATOR_JHIPSTER}@${version}`));
+                const generatorCommand = `${commandPrefix} ${GENERATOR_JHIPSTER}@${version} --dev --no-lockfile --ignore-scripts`;
+                this.log.info(generatorCommand);
```
here

deepu105 commented on generators/upgrade/index.js:
```diff
@@ -295,24 +306,40 @@ module.exports = UpgradeGenerator.extend({
         if (!this.skipInstall) {
-            shelljs.rm('-rf', 'node_modules');
             this.log('Installing dependencies, please wait...');
+            this.log.info('Removing the node_modules directory');
```
here

gmarziou replied:
BTW, here I don't think it really makes sense to delete `node_modules`, yarn is supposed to manage deps better and probably npm5 too. IMO, it just slows down the upgrade process.

deepu105 commented on generators/generator-base.js:
```diff
@@ -1683,6 +1684,24 @@ module.exports = class extends PrivateBase {
+     *
+     * @param {string} value - message to print
+     */
+    info(msg) {
```
the logging related methods are kind of spread around. There is debug in base-private. may be we should bring them all in this file also may be make it more clearer by making a wrapper object so that we can do
```
this.log.info()
this.log.debug()
this.log.error()
```
and so on. WDYT?

deepu105 replied:
since log is already defined by yeoman may be we should use `logger` namespace

gmarziou replied:
Yes I thought about it too but I did not want to impact too much code. From the log statuses provided by yeoman, I like `ok()` and `info()` and `maybe error()` (I don't remember it well) but I'm not sure the others are that useful, for instance I tried to use `invoke()` for all git calls but finally preferred `info()`. 

Also I must admit that I did not understand how their indentation system works with [emitted events](https://github.com/yeoman/environment/blob/master/lib/util/log.js#L51-L52).

gmarziou replied:
and I agree that we should have a uniform way to log success, error, info and debug rather than using all kind of colors with chalk. Yeoman's doc does not provide much guidance

gmarziou commented:
I rebased and made following changes:
- replaced this.log.info() by this.info()
- used commander for parsing new option `--prefer-local` which is useful when developing the upgrade generator itself with yarn link.

deepu105 commented on cli/jhipster.js:
```diff
@@ -35,21 +35,31 @@ if (!semver.satisfies(currentNodeVersion, minimumNodeVersion)) {
-requireCLI(preferGlobal);
+let preferLocal = true;
+
+program.command('upgrade')
```
doesnt this interfere with the original config in the cli file?

gmarziou replied:
I don't think so because I tested it (not all generators but upgrade and app are fine) and because program is defined as a const in both jhipster.js and cli.js so if they were interfering I would expect cli to raise an error for assigning twice a constant. I will check under debugger anyway.

gmarziou replied:
I was wrong: under debugger I was able to see that `program` is a global variable and in cl.js it is still holding the options set in jhipster.js. In fact, only the ref to the object is a const. I will rollback to previous code that did not use commander.

deepu105 replied:
ya I think that is safer

deepu105 commented on cli/jhipster.js:
```diff
@@ -35,21 +34,23 @@ if (!semver.satisfies(currentNodeVersion, minimumNodeVersion)) {
+
+// Don't use commander for parsing command line to avoid polluting it in cli.js 
+// --prefer-local: Always resolve node modules locally (useful when using linked module)
+if ((process.argv.indexOf('upgrade') > -1) && (process.argv.indexOf('--prefer-local') < 0)) {
```
you could also do `process.argv.includes('upgrade') && process.argv.includes('--prefer-local'))` which seems cleaner.

gmarziou replied:
Good idea, will do.
